packages/flutter/lib/src/material/text_field.dart

```diff
@@ -1077,6 +1110,24 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
         _showSelectionHandles = !widget.readOnly;
       }
     }
+
+    if (widget.statesController == oldWidget.statesController) {
+      final bool isOldWidgetEnabled = oldWidget.enabled ?? oldWidget.decoration?.enabled ?? true;
+
+      if (_isEnabled != isOldWidgetEnabled) {
+        _statesController.update(MaterialState.disabled, !_isEnabled);
+      }
+
+      _updateMaterialErrorState();
+    } else {
+      oldWidget.statesController?.removeListener(_handleStatesControllerChange);
+      if (widget.statesController != null) {
+        _internalStatesController?.dispose();
+        _internalStatesController = null;
+      }
+      _initStatesController();
+      _updateMaterialErrorState();
```

Comment:
consider always updating every state in `_stateController` here in `didUpdateWidget`. I think that will make it easier to read.